cm31_ntt/src/cm31.rs

```diff
@@ -0,0 +1,537 @@
+/// Complex M31 field arithmetic.
+ 
+use serde::{Serialize, Deserialize};
+use crate::rm31::{ RF, P };
+use std::ops::{ Add, AddAssign, Sub, SubAssign, Neg, Mul, MulAssign };
+use core::fmt::Display;
+use std::convert::{ From, Into };
+use num_traits::Zero;
+use num_traits::identities::One;
+use num_traits::pow::Pow;
+use rand::distributions::{Distribution, Standard};
+use rand::Rng;
+
+#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
+pub struct CF {
+    pub a: RF, // the real part
+    pub b: RF, // the imaginary part
+}
+
+// The 8th root of unity and its negation
+pub const W_8: CF = CF { 
+    a: RF { val: 0x00008000},
+    b: RF { val: 0x00008000 }
+};
+
+pub const W_8_NEG_1: CF = CF {
+    a: RF { val: 0x7fff7fff},
+    b: RF { val: 0x7fff7fff }
+};
+
+// The 4th root of unity and its negation
+pub const W_4: CF = CF {
+    a: RF { val: 0x00000000},
+    b: RF { val: 0x00000001 }
+};
+
+pub const W_4_NEG_1: CF = CF {
+    a: RF { val: 0x7fffffff},
+    b: RF { val: 0x7ffffffe }
+};
+
+
+/// Returns the 2nd to n-th roots of unity (inclusive).
+pub fn gen_roots_of_unity(n: usize) -> Vec<CF> {
+    assert!(n > 1);
+    let w2 = CF::root_of_unity_2();
+    let mut w = w2;
+    let mut res = vec![w2];
+    for _ in 2..n+1 {
+        w = w.try_sqrt().unwrap();
+        res.push(w);
+    }
+    res
+}
+
+impl CF {
+    pub fn new(real: u32, imag: u32) -> CF {
+        CF { a: RF::new(real), b: RF::new(imag) }
+    }
+
+    pub const fn real(self) -> RF {
+        self.a
+    }
+
+    pub const fn imag(self) -> RF {
+        self.b
+    }
+
+    pub fn mul_by_f(self, f: RF) -> CF {
+        CF { 
+            a: f * self.real(),
+            b: f * self.imag(),
+        }
+    }
+
+    pub fn try_inverse(&self) -> Option<Self> {
+        if self.a.val == 0 && self.b.val == 0 {
+            return None;
+        }
+
+        let a2b2 = (self.a * self.a + self.b * self.b).reduce();
+        if a2b2.is_zero() {
+            return None;
+        }
+
+        let a2b2_inv = a2b2.try_inverse().unwrap().reduce();
+        debug_assert!((a2b2 * a2b2_inv).reduce() == RF::new(1));
+
+        let neg_b = self.b.neg();
+        let a_neg_b = CF { a: self.a, b: neg_b };
+
+        let result = a_neg_b.mul_by_f(a2b2_inv);
+        Some(result)
+    }
+
+    pub fn reduce(self) -> CF {
+        CF { a: self.a.reduce(), b: self.b.reduce() }
+    }
+
+    pub fn root_of_unity_2() -> CF {
+        CF::new(0x7ffffffe, 0)
+    }
+
+    /// Returns the 4th root of unity. Since there are 2 options for this value, select the one you
+    /// want using the input `i`.
+    /// The 4th root of unity is (0, +-1)
+    /// The options denoted by `i` are:
+    /// 0. ( 0,  v)
+    /// 1. ( 0, -v)
+    pub fn root_of_unity_4(i: u32) -> Result<CF, String> {
```

Comment:
I recommend using the `anyhow` crate over plain `String` errors, which is what we do elsewhere in this codebase.